apps/lockfile-explorer-web/src/linter.ts

```diff
@@ -0,0 +1,238 @@
+import { LockfileEntry, LockfileEntryFilter } from './parsing/LockfileEntry';
+
+interface ILockfileEntryGroup {
+  entryName: string;
+  versions: LockfileEntry[];
+}
+
+export const linter = (entries: LockfileEntry[]) => {
+  const packageEntries = entries.filter((entry) => entry.kind === LockfileEntryFilter.Package);
+  const projectEntries = entries.filter((entry) => entry.kind === LockfileEntryFilter.Project);
+
+  // Get the list of side-by-side versions
+  const reducedEntries = packageEntries.reduce((groups: { [key in string]: LockfileEntry[] }, item) => {
+    const group = groups[item.entryPackageName] || [];
+    group.push(item);
+    groups[item.entryPackageName] = group;
+    return groups;
+  }, {});
+  let groupedEntries: ILockfileEntryGroup[] = [];
+  for (const [packageName, entries] of Object.entries(reducedEntries)) {
+    groupedEntries.push({
+      entryName: packageName,
+      versions: entries
+    });
+  }
+
+  const sideBySide = groupedEntries.filter((entry) => entry.versions.length > 1);
+  const sideBySidePackageNames = new Set(sideBySide.map((s) => s.entryName));
+
+  // set of package names that we need to log, calculated by going from project root downwards until we encounter a
+  // side by side version
+  const clusterNodes: {
+    [key in string]: {
+      duplicateVersions: Set<string>;
+      allowedConnected: {
+        [key in string]: string[][];
+      };
+    };
+  } = {};
+
+  // Mapping of projectName -> dep.name -> set of versions that it has
+  const connectedProjectsToClusters: { [key in string]: { [key in string]: Set<string> } } = {};
+
+  for (const projectEntry of projectEntries) {
+    const visited = new Set<LockfileEntry>();
+    connectedProjectsToClusters[projectEntry.entryPackageName] = {};
+
+    // This keeps track of any side-by-side dependencies we encounter in this project
+    // We save only one version because if we encounter any other versions for the same side-by-side dependency in the same project
+    // Then we know that this project contains more than one side-by-side dependency in it's tree
+    const seenSideBySide: { [key in string]: string } = {};
+    // Keep track of the lockfile entries we visit when traversing the graph
+    const parentStack: string[] = [];
+
+    function helper(currNode: LockfileEntry) {
```

Comment:
Use a more descriptive name for this function.

Reply:
fixed